### ISSUE TITLE
Health Check

### ISSUE THREAD

**Issue by vbaidak** [offen/docker-volume-backup#384]
Hi Team,

Is there any health check command?

**Comment by m90** [comment]
What's the failure scenario you would like to check for with such a health check? Right now, as long as the container did not exit, things are assumed running as configured (which obviously might not be true).

However, if this image would implement a health check, I wouldn't even know which conditions to check for. Do you have something specific in mind?

**Comment by vbaidak** [comment]
Yeah, that's why I'm asking. I'm ok with just having `unless-stopped` condition and asking only for the case if there is any process inside a container that is running and could potentially fail without container fail, so that we are not getting backups and do not know about this

**Comment by m90** [comment]
The process in the container works kind of like `crond` and is designed to never ever fail and instead just run the scheduled jobs, no matter if these invocations fail or not. This means the container being up does not tell you anything about whether backups have successfully been performed or not, the only guarantee you get is that it has been tried. I.e. you could have a container that is up for over a year but hasn't performed a single backup as you have accidentally provided bad credentials for your storage backend.

If you want job-level visibility instead, you can either monitor your logs or use the notification feature, which can notify you about either all runs or just failed ones. Docs are here https://offen.github.io/docker-volume-backup/how-tos/set-up-notifications.html

There are also ideas about exposing metrics via HTTP (e.g. Prometheus), but I don't know how / if this will ever happen: https://github.com/offen/docker-volume-backup/issues/64

**Comment by vbaidak** [comment]
Thanks for the explanation!
Well, I've already configured email notifications and they work well, so my backup failures will be covered.
What about the process itself? Maybe we can at least check if it is running via `ps` or does it have PID 1?

**Comment by m90** [comment]
> What about the process itself?

This should be the responsibility of Docker or whatever tool runs the container in that case. If the process exits, the container does so as well, and in this case whoever is supervising the container should be taking care of whether it wants to restart the container or not.

**Comment by vbaidak** [comment]
I see

**Comment by huyz** [comment]
Maybe a healthcheck would be useful for this scenario:

Occasionally, when I trigger `backup`, I get this:
```
time=2024-12-14T07:12:24.192Z level=ERROR msg="Fatal error running command: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?" error="main.(*command).runAsCommand: error running script: main.runScript.func4: error running script: main.runScript.func4.1.(*script).withLabeledCommands.2: error running archive-pre commands: main.(*script).runLabeledCommands: error querying for containers: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?"
```

I don't know why this is happening on my Docker Desktop for Windows, when my other services are running fine.

Once, I restart the container, all is good again.

I do use `NOTIFICATION_URLS` in conjunction with healthchecks.io, so I am notified when backups don't work.
But if there were a healthcheck, then my cron job, which restarts unhealthy containers, could automatically resolve the issue without a manual intervention from me.

**Comment by m90** [comment]
Currently there's a whole class of errors that aren't caught by the notification feature (bascially everything that goes wrong before the notification feature is initialized). I would think rectifying this would be the proper way to solve your issue.

**Comment by QarthO** [comment]
A healthcheck can be useful for testing that the backup location and endpoint is accessible.

For example, lets say an S3 bucket is no longer accessible, or the local drive path is no longer available, then there'd be an issue and the backup wouldn't work.

**Comment by m90** [comment]
@QarthO this kind of error scenario can be monitored using the notification feature

**Comment by QarthO** [comment]
Yes, a notification after the backup fails, not prior. Which is why a healthcheck would be perfect for this

**Comment by m90** [comment]
I'd think permanent monitoring of arbitrary infrastructure should better be done using different tools and is out of scope for this one.
